225 move ai part from java to kotlin

Mohammad-Masoomi-Homayoun opened the pull request.

mohammadmasoomih committed:
Rename .java to .kt

mohammadmasoomih committed:
Converting java files into kotlin

Mohammad-Masoomi-Homayoun commented:
please review @sharifyy

sharifyy commented on src/main/kotlin/com/linchpino/ai/service/impl/RoadmapServiceImpl.kt:
```diff
@@ -0,0 +1,32 @@
+        return applicationContext.getBean(serviceName.componentName, AIService::class.java)
+    }
+
+    override fun getRoadmap(targetLevel: String, resumeFile: File?): String? {
```
targetLevel is not a nullable type therefore null check is redundant.
based  on the code written in the controller(the only client), client always passes resumeFile, so there is no need to define resumeFile as nullable and the null check is redundant

sharifyy commented on src/main/kotlin/com/linchpino/ai/model/RequestDetail.kt:
```diff
@@ -0,0 +1,3 @@
+package com.linchpino.ai.model
+
+class RequestDetail(var targetLevel: String, var resume: Resume)
```
could be a data class with final fields

sharifyy commented on src/main/kotlin/com/linchpino/ai/service/impl/ResumeService.kt:
```diff
@@ -0,0 +1,72 @@
+    @Value("\${spring.ai.roadmap-attempt-limit}")
+    private val roadmapAttemptLimit = 0
+
+    fun create(file: File?): Resume {
```
Do not pass null file and remove the null check.

sharifyy commented on src/main/kotlin/com/linchpino/ai/service/impl/ResumeService.kt:
```diff
@@ -0,0 +1,72 @@
+        return Resume(email, lines)
+    }
+
+    fun save(resume: Resume): Resume {
```
resume is not nulll and the null check is redundant

sharifyy commented on src/main/kotlin/com/linchpino/ai/service/impl/ResumeService.kt:
```diff
@@ -0,0 +1,72 @@
+    fun create(file: File?): Resume {
+        requireNotNull(file) { "File cannot be null" }
+        val lines = readLines(file)
+        val email = Resume.findEmail(lines)
```
Resume.findEmail delagates to Contact.findEmail, since both are public and static you should directly call Contact.findEmail

Mohammad-Masoomi-Homayoun replied:
You’re right; I thought that repeatedly calling resume to locate the email might improve readability.

sharifyy commented on src/main/kotlin/com/linchpino/ai/service/impl/ResumeService.kt:
```diff
@@ -0,0 +1,72 @@
+    fun save(resume: Resume): Resume {
+        Objects.requireNonNull(resume, "Resume cannot be null")
+        Objects.requireNonNull(resume.email, "Email cannot be null")
+        return resumeRepository.save(resume)
```
if resume lines contain not utf-8 characters database throws exception

sharifyy commented on src/main/kotlin/com/linchpino/ai/service/impl/ResumeService.kt:
```diff
@@ -0,0 +1,72 @@
+        return resumeRepository.save(resume)
+    }
+
+    fun findById(id: Long): Resume? {
```
Never used.

sharifyy commented on src/main/kotlin/com/linchpino/ai/service/impl/ResumeService.kt:
```diff
@@ -0,0 +1,72 @@
+        return resumeRepository.findById(id).orElse(null)
+    }
+
+    fun findAll(): MutableList<Resume?> {
```
Never used

sharifyy commented on src/main/kotlin/com/linchpino/ai/service/impl/ResumeService.kt:
```diff
@@ -0,0 +1,72 @@
+            Loader.loadPDF(file).use { document ->
+                logger.info("Reading the file: {}", file.name)
+                val textStripper = PDFTextStripper()
+                return ArrayList(
```
call to ArrayList constructor is redundant

sharifyy commented on src/main/kotlin/com/linchpino/ai/service/impl/ResumeService.kt:
```diff
@@ -0,0 +1,72 @@
+                val textStripper = PDFTextStripper()
+                return ArrayList(
+                    Arrays.asList(
+                        *textStripper.getText(document).split("\n".toRegex()).dropLastWhile { it.isEmpty() }
```
split returns list. there is no need to bring Java's verbosity to Kotlin. No spreading and converting to byte array and again converting bytearray to list is needed.
you can just: 
return PDFTextStripper().getText(document).split("\n".toRegex()).dropLastWhile { it.isEmpty() }

sharifyy commented on src/main/kotlin/com/linchpino/ai/model/AIServiceName.kt:
```diff
@@ -0,0 +1,32 @@
+import com.linchpino.ai.service.impl.ChatGPTServiceImpl
+import com.linchpino.ai.service.impl.GeminiServiceImpl
+
+enum class AIServiceName(val label: String, @JvmField val componentName: String, private val isDefault: Boolean) {
```
this enum is redundant. You can reach the same functionality with one line of code

Mohammad-Masoomi-Homayoun replied:
I created this enum to enhance code extensibility.

sharifyy commented on src/main/kotlin/com/linchpino/ai/model/AIServiceName.kt:
```diff
@@ -0,0 +1,32 @@
+import com.linchpino.ai.service.impl.ChatGPTServiceImpl
+import com.linchpino.ai.service.impl.GeminiServiceImpl
+
+enum class AIServiceName(val label: String, @JvmField val componentName: String, private val isDefault: Boolean) {
```
this enum is redundant. You can reach the same functionality with one line of code

Mohammad-Masoomi-Homayoun replied:
same https://github.com/linchpino/linchpino-core/pull/147#discussion_r1808183204

sharifyy commented on src/main/kotlin/com/linchpino/ai/service/impl/RoadmapServiceImpl.kt:
```diff
@@ -0,0 +1,32 @@
+        val requestDetail = RequestDetail(targetLevel, resumeService.save(resume))
+        return loadAIService(serviceName).talkToAI(requestDetail)
+    }
+}
```
you can easily remove AIServiceName enum and replace it with a list. here is my suggestion:

@Service
class RoadmapServiceImpl(private val applicationContext: ApplicationContext, private val resumeService: ResumeService) :
    RoadmapService {
    @Value("\${spring.ai.default-service-provider}")
    private val aiServiceName: String? = null

    private val aiServices = listOf(ChatGPTServiceImpl.COMPONENT_NAME,GeminiServiceImpl.COMPONENT_NAME)

    override fun getRoadmap(targetLevel: String, resumeFile: File?): String? {
        val serviceName = aiServices.firstOrNull { it == aiServiceName } ?: GeminiServiceImpl.COMPONENT_NAME
        val resume = resumeService.create(resumeFile)
        val requestDetail = RequestDetail(targetLevel, resumeService.save(resume))
        return applicationContext.getBean(serviceName, AIService::class.java).talkToAI(requestDetail)
    }
}

sharifyy commented on src/main/kotlin/com/linchpino/ai/service/AIService.kt:
```diff
@@ -0,0 +1,7 @@
+import com.linchpino.ai.model.RequestDetail
+
+interface AIService {
+    fun talkToAI(requestDetail: RequestDetail?): String?
```
Do not propagate null into your software. make requestDetail a non nullable type

sharifyy commented on src/main/kotlin/com/linchpino/ai/model/Prompt.kt:
```diff
@@ -27,26 +34,13 @@ public class Prompt {
-        if (requestDetail == null) {
-            throw new IllegalArgumentException("Request detail cannot be null");
+        @JvmStatic
+        fun of(requestDetail: RequestDetail?): Prompt {
```
this method does nothing other than checking for null, and you should avoid passing null in the first place. you can easily remove this method

sharifyy commented on src/main/kotlin/com/linchpino/ai/model/Prompt.kt:
```diff
@@ -1,8 +1,15 @@
 
-    public static final String ROADMAP_PROMPT = """
+    companion object {
+        val ROADMAP_PROMPT: String = """
```
there is no need for a companion object here. define this string as instance member. if you want you can use top level members too.
private const val Roadmap_Prompt = "example"

sharifyy commented on src/main/kotlin/com/linchpino/ai/model/Roadmap.kt:
```diff
@@ -0,0 +1,20 @@
+package com.linchpino.ai.model
+
+class Roadmap {
```
Never used

mohammadmasoomih committed:
Resolve findings

mohammadmasoomih committed:
Resolve findings
- Remove unnecessary null point checkers
- Introduced Resume.removeInvalidUtf8Bytes method to remove the invalid characters to save
